src/Bicep.LangServer/Completions/ParamsCompletionContext.cs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bicep.Core;
+using Bicep.Core.Extensions;
+using Bicep.Core.Navigation;
+using Bicep.Core.Parsing;
+using Bicep.Core.Syntax;
+using Bicep.Core.Workspaces;
+using Bicep.LanguageServer.CompilationManager;
+using Bicep.LanguageServer.Extensions;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Bicep.LanguageServer.Completions
+{
+    public class ParamsCompletionContext
+    {   
+        // completions will replace only these token types
+        // all others will result in an insertion upon completion commit
+        private static readonly ImmutableHashSet<TokenType> ReplaceableTokens = new[]
+        {
+            TokenType.Identifier,
+            TokenType.Integer,
+            TokenType.StringComplete
+        }.Concat(LanguageConstants.Keywords.Values).ToImmutableHashSet();
+
+        public ParamsCompletionContextKind Kind { get; }
+
+        public Range ReplacementRange { get; }
+
+        public ParamsCompletionContext(ParamsCompletionContextKind kind, Range replacementRange)
+        {
+            Kind = kind;
+            ReplacementRange = replacementRange;
+        }
+
+        public static ParamsCompletionContext Create(ParamsCompilationContext paramsCompilationContext, int offset)
+        {
+            var paramsFile = paramsCompilationContext.ParamsSemanticModel.bicepParamFile;
+            var matchingNodes = SyntaxMatcher.FindNodesMatchingOffset(paramsFile.ProgramSyntax, offset);
+            if (!matchingNodes.Any())
+            {
+                // this indicates a bug
+                throw new ArgumentException($"The specified offset {offset} is outside the span of the specified {nameof(ProgramSyntax)} node.");
+            }
+            var replacementRange = GetReplacementRange(paramsFile, matchingNodes[^1], offset);
+            
+            var topLevelDeclarationInfo = SyntaxMatcher.FindLastNodeOfType<ITopLevelNamedDeclarationSyntax, SyntaxBase>(matchingNodes);
+
+
+            var kind = ConvertFlag(IsParamAssignmentContext(matchingNodes, offset), ParamsCompletionContextKind.ParamAssignment);
+
+                    //    ConvertFlag(IsTopLevelDeclarationStartContext(matchingNodes, offset), BicepCompletionContextKind.TopLevelDeclarationStart) |
```

Comment:
Should I worry about TopLevelDeclarations (specifically for keywords)?

Reply:
That can be added in a future PR, but it should be covered here.